### PR TITLE
Added more automatic generator functions (#694)

### PR DESCRIPTION
**Pull request by MatteoVoges**
Fixes issue #694

## Proposed Changes
**New generic `random` function:**
Generates a random string based on given parameters.
### Usage
```yaml
secret: ?{ref:path/to/secret||random:type:nchars:special}
```
The first parameter `type` has to be one of:
* `str` : generator function for alphanumeric characters, will be url-token-safe
* `int` : generator function for digits (0-9)
* `loweralpha` : generator function for lowercase letters (a-z)
* `upperalpha` : generator function for uppercase letters (A-Z)
* `loweralphanum` : generator function for lowercase letters and numbers (a-z and 0-9)
* `upperalphanum` : generator function for uppercase letters and numbers (A-Z and 0-9)
* `special` : generator function for alphanumeric characters and given special characters

I'm always open for name suggestions.

All these types have in common, that the second parameter `nchars` indicates how many characters the generated secret contains.
The default value for this is
  * `random:str` : 43 characters
  * `random:int` : 16 characters
  * other : 8 characters

The `special` type has a third argument that indicates which special characters are allowed for the generation.
The default for that is `string.punctuation` (``!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~``).

**NOTE:** in manual usage the characters `|` and `:` are used for function and argument parsing, so they can't be specified.

This might be a 
## Breaking change
To pass special characters into the generator pool, we have to change the **REF_TOKEN_TAG_PATTERN**-regex.
The new regex `(\?{(\w+:[\w\-\.\@\=\/\:]+)(\|(?:(?:\|\w+)(?::\S*)*)+)?\=*})` matches all existing use cases, but is more restricted.
So be beware of unexpected errors.

Furthermore the existing generator functions `randomstr` and `lowernumalpha` are still working and still behave the same.
But they will be removed soon.

## Examples
Here are some example cases with expected output:
* Tag: `?{ref:path/to/secret||random:int}` generates a secret containing 16 (default) digits, i.e. `1254872907909324`
* Tag: `?{ref:path/to/secret||random:upperalphanum:12}` generates a secret containing 12 chars from uppercase letters and numbers, i.e. `VUPA0GIO47C2`
* Tag: `?{ref:path/to/secret||random:special:16:!$%&}` generates a secret containing 16 chars from alphanumeric letters and given special characters `!$%&`, i.e. `!kQT&m%csSDsdaBy`

You could leave the argument empty, so `?{ref:path/to/secret||random:special:12:}` generates a alphanumeric string without any special characters.

## Documentation and testing
* Updated Tests
* **TBD:** Update docs for latest iteration.

## Minor Issue
tl;dr: Some errors don't get caught and could produce some unexpected behavior.
See detailed: https://github.com/kapicorp/kapitan/issues/694#issuecomment-1277919468

### CODE REVIEW

File: kapitan/refs/functions.py [hunk 1]
```diff
@@ -20,20 +20,30 @@
...
+        "loweralphanum": lower_alpha_num,
+        "upperalpha": upper_alpha,
+        "upperalphanum": upper_alpha_num,
+        "alphanumspec": alphanumspec,
     }
 
-    return func_lookup[func_name](ctx, *func_params)
-
 
 def randomstr(ctx, nbytes=""):
```
Comment (MatteoVoges) [inline kapitan/refs/functions.py:48]: I could update `randomstr` to a more generic solution as well.
It would look like:
```python
def randomstr(ctx, nchars="43"):
    pool = string.ascii_letters + string.digits + "-_"
    generic_alphanum(ctx, nchars, pool)
```
Also we could use `alphanumspec` for that:
```python
def randomstr(ctx, nchars="43"):
    alphanumspec(ctx, nchars, "-_")
```

File: tests/test_refs.py [hunk 2]
```diff
@@ -440,3 +441,113 @@ def test_ref_function_loweralphanum(self):
...
+        # Test with every allowed special char, nchars=32
+        allowed_special_chars = set(string.punctuation).difference(":|")
+        for special_char in allowed_special_chars:
+            tag = "?{plain:ref/alphanumspec||alphanumspec:32:" + special_char + "}"
+            REF_CONTROLLER[tag] = RefParams()
+            REVEALER._reveal_tag_without_subvar.cache_clear()
+            revealed = REVEALER.reveal_raw_file(file_with_tags)
+            self.assertEqual(len(revealed), 32)
+            intersection = set(string.punctuation).intersection(revealed)
+            self.assertTrue(intersection.issubset(set(special_char)))
```
Comment (MatteoVoges) [inline tests/test_refs.py:553]: I think the for loop is unneccessary here. We could test it directly with `allowed_special_chars` as one string as input.

**Comment by ademariag** [comment]
suggestion from Kapitan Workgroup:

introduce one new function `random` with the following parameters

`random:<class>:<length>`

where class is one of:

`str`: same as currently using `randomstr`
`int` : generator function for digits (0-9)
`loweralpha` : generator function for lowercase letters (a-z)
`upperalpha` : generator function for uppercase letters (A-Z)
`upperalphanum` : generator function for uppercase letters and numbers (A-Z and 0-9)
`alphanumspec` : generator function for alphanumeric characters and given special characters

i.e. 
`random:str` produces same results as current `randomstr`
`random:int` produces a random int
`random:loweralpha:15` produces a 15 character `loweralpha` string

File: kapitan/refs/functions.py (hunk 1 above)
Comment (MatteoVoges) [reply kapitan/refs/functions.py:44]: Fixed with latest iteration

File: tests/test_refs.py (hunk 2 above)
Comment (MatteoVoges) [reply tests/test_refs.py]: Fixed with latest iteration

**Comment by MatteoVoges** [comment]
Do you think in `test_refs.py` every single option for `type` should be checked? Or is it enough to take some samples?
